src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackEncoder.cs

```diff
@@ -502,56 +518,87 @@ public static bool EncodeDynamicTableSizeUpdate(int value, Span<byte> destinatio
 
                 // Calculate length of all parts and separators.
-                foreach (string part in values)
+                if (valueEncoding is null || valueEncoding.IsSingleByte)
                 {
-                    valueLength = checked((int)(valueLength + part.Length));
+                    valueLength = checked((int)(values.Length - 1) * separator.Length);
+                    foreach (string part in values)
+                    {
+                        valueLength = checked((int)(valueLength + part.Length));
+                    }
+                }
+                else
+                {
+                    valueLength = checked((int)(values.Length - 1) * valueEncoding.GetByteCount(separator));
+                    foreach (string part in values)
+                    {
+                        valueLength = checked((int)(valueLength + valueEncoding.GetByteCount(part)));
+                    }
                 }
-
-                Debug.Assert(separator != null);
-                valueLength = checked((int)(valueLength + (values.Length - 1) * separator.Length));
 
                 destination[0] = 0;
                 if (IntegerEncoder.Encode(valueLength, 7, destination, out int integerLength))
                 {
                     Debug.Assert(integerLength >= 1);
-
-                    int encodedLength = 0;
-                    for (int j = 0; j < values.Length; j++)
+                    destination = destination.Slice(integerLength);
+                    if (destination.Length >= valueLength)
                     {
-                        if (j != 0 && !EncodeStringLiteralValue(separator, destination.Slice(integerLength), out encodedLength))
+                        if (valueEncoding is null)
                         {
-                            return false;
+                            for (int i = 0; i < values.Length; i++)
+                            {
+                                if (i != 0)
+                                {
+                                    EncodeValueStringPart(separator, destination);
+                                    destination = destination.Slice(separator.Length);
+                                }
+
+                                string value = values[i];
+                                EncodeValueStringPart(value, destination);
+                                destination = destination.Slice(value.Length);
+                            }
                         }
-
-                        integerLength += encodedLength;
-
-                        if (!EncodeStringLiteralValue(values[j], destination.Slice(integerLength), out encodedLength))
+                        else
                         {
-                            return false;
+                            int written;
+                            for (int i = 0; i < values.Length; i++)
+                            {
+                                if (i != 0)
+                                {
```

Comment:
I'm sure it won't make any difference due to encoding overhead but this branch could be removed by writing [0] first and then `for(int i = 1; ...`

Reply:
Moved the `i == 0` case outside (HPack and QPack)